src/ProgressLogging.jl

```diff
@@ -3,13 +3,168 @@ module ProgressLogging
 # Use README as the docstring of the module:
 @doc read(joinpath(dirname(@__DIR__), "README.md"), String) ProgressLogging
 
-export @progress, @withprogress, @logprogress
+export @progress, @progressid, @withprogress, @logprogress
 
 using Base.Meta: isexpr
+using UUIDs: UUID, uuid4, uuid5
 using Logging: Logging, @logmsg, LogLevel
 
 const ProgressLevel = LogLevel(-1)
 
+"""
+    ProgressLogging.ROOTID
+
+This is used as `parentid` of root [`Progress`](@ref)es.
+"""
+const ROOTID = UUID(0)
+
+"""
+    ProgressLogging.Progress(id, [fraction]; [parentid, name, done])
+
+# Usage: Progress log record provider
+
+Progress log record can be created by using the following pattern
+
+```julia
+id = uuid4()
+try
+    @info Progress(id)  # create a progress bar
+    # some time consuming task
+    # ...
+    @info Progress(id, 0.1)  # update progress to 10%
+    # ...
+finally
+    @info Progress(id, done = true)  # close the progress bar
+end
+```
+
+It is recommended to use [`@withprogress`](@ref),
+[`@logprogress`](@ref), and optionally [`@progressid`](@ref) to create
+log records.
+
+# Usage: Progress log record consumer (aka progress monitor)
+
+It is recommended to use [`ProgressLogging.asprogress`](@ref) instead
+of checking `message isa Progress`.  Progress monitors can retrieve
+progress-related information from the following properties.
+
+# Properties
+- `fraction::Float64`: it can take following values:
+  - `0 <= fraction < 1`
+  - `fraction = NaN`: indeterminate progress
+  - `fraction >= 1`: completed
+- `id::UUID`: Identifier of the task whose progress is at `fraction`.
+- `parentid::UUID`: The ID of the parent progress.  It is set to
+  [`ProgressLogging.ROOTID`](@ref) when there is no parent progress.
+  This is used for representing progresses of nested tasks.  Note that
+  sub-tasks may be executed concurrently; i.e., there can be multiple
+  child tasks for one parent task.
+- `name::String`: Name of the progress bar.
+- `done::Bool`: `true` if the task is done.
+"""
+Base.@kwdef struct Progress
+    id::UUID
+    parentid::UUID = ROOTID  # not nested by default
+    fraction::Float64 = NaN
+    name::String = ""
+    done::Bool = false
+end
+
+Progress(id::UUID, fraction::Real = NaN; kwargs...) =
+    Progress(; kwargs..., fraction = fraction, id = id)
+
+# Define `string`/`print` so that progress log records are (somewhat)
+# readable even without specific log monitors.
+function Base.print(io::IO, progress::Progress)
```

Comment:
Shouldn't this be overloading `Base.show`? `print` falls back to two-arg `show` anyways.

Reply:
Thanks for the review!

I'd like to avoid overloading `show(::IO, ::Progress)` since that's used from `repr(::Progress)` which ideally should produce a valid Julia expression.

I think it is OK to define `print` occasionally. The documentation says:

> `print` falls back to calling `show`, so most types should just define `show`. **Define `print` if your type has a separate "plain" representation.** For example, `show` displays strings with quotes, and `print` displays strings without quotes.
>
> --- https://docs.julialang.org/en/latest/base/io-network/#Base.print

I'm overloading `print` to control `string(message)` as that's how the message is processed:

* [`handle_message(::ConsoleLogger, ...)`](https://github.com/JuliaLang/julia/blob/9b7ea2f0b3bf4f1062eb2cee86f3cfcb197e54eb/stdlib/Logging/src/ConsoleLogger.jl#L108)
* [`handle_message(::SimpleLogger, ...)`](https://github.com/JuliaLang/julia/blob/9b7ea2f0b3bf4f1062eb2cee86f3cfcb197e54eb/base/logging.jl#L546)

However, I just noticed that the Logging documentation says:

> The *message*  is an object describing the event. By convention `AbstractString`s passed as messages are assumed to be in markdown format. Other types will be displayed using `show(io,mime,obj)` according to the display capabilities of the installed logger.
>
> --- https://docs.julialang.org/en/latest/stdlib/Logging/

So maybe it also makes sense to define `show(::IO, ::MIME"text/plain", ::Progress)`?  But this could be confusing as we'd have

```julia
julia> Progress(uuid4(), 0.1)
Progress: 10%
```

@c42f Maybe we should "fix" the Logging documentation to something like this?

> The *message*  is an object describing the event. By convention `AbstractString`s passed as messages are assumed to be in markdown format. Other types will be displayed using **`print(io, obj)` or `string(obj)` for text-based logger and possibly `show(io,mime,obj)` for other "rich" display types used in the installed logger.**

Reply:
Good point. Looks good as-is then.

Reply:
Sure, I'm open to changing the `Base` docs for Logging — that looks like a sensible convention for the message. IMHO the various `print` vs `show` with and without MIME info are pretty confusing...